runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -796,14 +797,69 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
          isStoreValueVT = isValue(storeValueConstraint);
          }
 
-      // If the array's component type is definitely not a value type, or if the value
-      // being assigned in an array store operation is definitely not a value type, add
-      // a delayed transformation to replace the helper call with inline code to
-      // perform the array element access.
-      //
-      if ((arrayConstraint != NULL && isCompTypeVT == TR_no)
-          || (isStoreFlattenableArrayElement && isStoreValueVT == TR_no))
+      bool canTransformFlattenedArrayElementLoadStore = false;
+      // If the array element is VT and flattened, add a delayed transformation to replace
+      // the helper call with sym refs access.
+      if ((isLoadFlattenableArrayElement || (isStoreFlattenableArrayElement && storeValueConstraint)) &&
+          arrayConstraint &&
+          isArrayElementFlattened(arrayConstraint) == TR_yes)
          {
+         TR_OpaqueClassBlock *arrayComponentClass = fe()->getComponentClassFromArrayClass(arrayConstraint->getClass());
+
+         if (arrayComponentClass)
+            {
+            if (isLoadFlattenableArrayElement)
+               {
+               canTransformFlattenedArrayElementLoadStore = true;
+               }
+            else // isStoreFlattenableArrayElement
+               {
+               TR_OpaqueClassBlock *valueClass = storeValueConstraint->getClass();
+               TR_YesNoMaybe isInstanceOfComponentType = comp()->fej9()->isInstanceOf(valueClass,
+                                                                                   arrayComponentClass,
+                                                                                   storeValueConstraint->isFixedClass(),
+                                                                                   arrayConstraint->isFixedClass());
+
+               // If the value that's being stored is not the same as the array component type, use helper call.
+               canTransformFlattenedArrayElementLoadStore = (isInstanceOfComponentType == TR_yes);
+               }
+            }
+         }
+
+      // If the array is known to be of a value type that is not flattened, transform the helper call to regular aaload and aastore.
+      bool canTransformNotFlattenedVTArrayElementLoadStore = false;
```

Comment:
A really minor suggestion - may I ask you to change the name of this variable to `canTransformUnflattenedVTArrayElementLoadStore`?

Reply:
Addressed in bdab531